src/libstd/keyword_docs.rs

```diff
@@ -1651,9 +1651,118 @@ mod use_keyword {}
 //
 /// Add constraints that must be upheld to use an item.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// `where` allows specifying constraints on lifetime and generic parameters.
+/// The [RFC] introducing `where` contains detailed informations about the
+/// keyword.
 ///
-/// [not yet complete]: https://github.com/rust-lang/rust/issues/34601
+/// # Examples
+///
+/// `where` can be used for constraints with traits:
+///
+/// ```rust
+/// fn new<T: Default>() -> T {
+///     T::default()
+/// }
+///
+/// fn new_where<T>() -> T
+/// where
+///     T: Default,
+/// {
+///     T::default()
+/// }
+///
+/// assert_eq!(0.0, new());
+/// assert_eq!(0.0, new_where());
+///
+/// assert_eq!(0, new());
+/// assert_eq!(0, new_where());
+/// ```
+///
+/// `where` can also be used for lifetimes.
+///
+/// This compiles because the lifetime of `longer` is superior to the lifetime
+/// of `shorter`, thus the constraint is respected:
+///
+/// ```rust
+/// fn select_where<'a, 'b>(s1: &'a str, s2: &'b str, second: bool) -> &'a str
+/// where
+///     'b: 'a,
+/// {
+///     if second {
+///         s2
+///     } else {
+///         s1
+///     }
+/// }
+///
+/// let outer = String::from("Long living ref");
+/// let longer = &outer;
+/// {
+///     let inner = String::from("Short living ref");
+///     let shorter = &inner;
+///
+///     assert_eq!(select_where(shorter, longer, false), shorter);
+///     assert_eq!(select_where(shorter, longer, true), longer);
+/// }
+/// ```
+///
+/// On the other hand, this will not compile: `shorter` does not have a lifetime
+/// that respects the constraint imposed by the `select_where` functions.
+///
+/// ```rust,compile_fail,E0597
+/// # fn select_where<'a, 'b>(s1: &'a str, s2: &'b str, second: bool) -> &'a str
+/// # where
+/// #     'b: 'a,
+/// # {
+/// #     if second {
+/// #         s2
+/// #     } else {
+/// #         s1
+/// #     }
+/// # }
+/// let outer = String::from("Long living ref");
+/// let longer = &outer;
+/// let res;
+/// {
+///     let inner = String::from("Short living ref");
+///     let shorter = &inner;
+///
+///     res = select_where(longer, shorter, false);
```

Comment:
So I went down a rabbit hole here - it's true that this fails to compile, but it's not because of the `where` clause. Rather, it's because `inner` was dropped at the end of the block, so it can't be used outside; that error would happen with or without the `where`. I tried removing the `assert_eq!()` afterwards and found that actually compiles without a problem, even though `outer` lives longer than `inner`! https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0e254df4b5f8ce301b2c253acbd04e94

I think what's going on is that `longer` is being implicitly reborrowed for a lifetime that's less than `inner`, something like `select_where(&*longer, shorter, false);`. Rather than try and mess with this more, would it be simpler to show what happens if you don't have the bound? I think that error makes it more clear why it's useful: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=37482da9175c6a0ae0615c3a8cc36f97

```
error[E0623]: lifetime mismatch
 --> src/main.rs:3:7
  |
1 | fn select_where<'a, 'b>(s1: &'a str, s2: &'b str, second: bool) -> &'a str {
  |                                          -------                   -------
  |                                          |
  |                                          this parameter and the return type are declared with different lifetimes...
2 |   if second {
3 |       s2
  |       ^^ ...but data from `s2` is returned here

error: aborting due to previous error
```

Reply:
That's both shorter and simpler, thanks !